File: src/commands/doctor.ts
import pc from "picocolors";
import { readJsonFile } from "../config/json.js";
import * as toml from "../config/toml.js";
import type { AgentId, RunOpts, ToolId } from "../registry.js";
import {
  ALL_AGENTS,
  ALL_TOOLS,
  detectAgent,
  toolHealthCheck,
  toolInstalledVersion,
  toolLatestVersion,
  toolRepair,
  verifyTool,
} from "../registry.js";
import * as colors from "../util/colors.js";
import type { HealthStatus } from "../util/health.js";
import { mcpSpawnHealthy } from "../util/mcpspawn.js";
import { formatPathFixResult, selfHealPath } from "../util/pathfix.js";
import * as paths from "../util/paths.js";
import { isUpToDate } from "../util/version.js";

function checkAgentMcpHealth(agent: AgentId, tool: ToolId): { healthy: boolean; reason?: string } {
  try {
    switch (agent) {
      case "claude": {
        const cfg = (readJsonFile(paths.claudePaths().globalJson) as Record<string, unknown>) ?? {};
        const srv = (cfg.mcpServers as Record<string, { command?: string; args?: string[] }>) ?? {};
        const entry = srv[tool];
        if (!entry) return { healthy: false, reason: "not configured" };
        return mcpSpawnHealthy(entry.command ?? "", entry.args ?? []);
      }
      case "opencode": {
        const cfg = (readJsonFile(paths.opencodePaths().config) as Record<string, unknown>) ?? {};
        const mcp =
          (cfg.mcp as Record<string, { type?: string; command?: string | string[] }>) ?? {};
        const entry = mcp[tool];
        if (!entry) {
          if (tool === "context-mode") {
            const plugins = (cfg.plugin as string[]) ?? [];
            if (
              plugins.some(
                (p) =>
                  typeof p === "string" && (p === "context-mode" || p.startsWith("context-mode@")),
              )
            ) {
              return { healthy: true };
            }
          }
          return { healthy: false, reason: "not configured" };
        }
        if (Array.isArray(entry.command)) {
          return mcpSpawnHealthy(entry.command[0] ?? "", entry.command.slice(1));
        }
        return mcpSpawnHealthy(entry.command ?? "");
      }
      case "codex": {
        const doc = toml.readTomlFile(paths.codexPaths().config);
        const mcpServers =
          (doc.mcp_servers as Record<string, { command?: string; args?: string[] }>) ?? {};
        const entry = mcpServers[tool];
        if (!entry) return { healthy: false, reason: "not configured" };
        return mcpSpawnHealthy(entry.command ?? "", entry.args ?? []);
      }
      case "copilot": {
        const cfg = (readJsonFile(paths.copilotPaths().mcpConfig) as Record<string, unknown>) ?? {};
        const srv = (cfg.mcpServers as Record<string, { command?: string; args?: string[] }>) ?? {};
        const entry = srv[tool];
        if (!entry) return { healthy: false, reason: "not configured" };
        return mcpSpawnHealthy(entry.command ?? "", entry.args ?? []);
      }
      case "droid": {
        const cfg = (readJsonFile(paths.droidPaths().mcpConfig) as Record<string, unknown>) ?? {};
        const srv = (cfg.mcpServers as Record<string, { command?: string; args?: string[] }>) ?? {};
        const entry = srv[tool];
        if (!entry) return { healthy: false, reason: "not configured" };
        return mcpSpawnHealthy(entry.command ?? "", entry.args ?? []);
      }
      case "devin": {
        const cfg = (readJsonFile(paths.devinPaths().mcpConfig) as Record<string, unknown>) ?? {};
        const srv = (cfg.mcpServers as Record<string, { command?: string; args?: string[] }>) ?? {};
        const entry = srv[tool];
        if (!entry) return { healthy: false, reason: "not configured" };
        return mcpSpawnHealthy(entry.command ?? "", entry.args ?? []);
      }
      case "antigravity": {
        for (const f of paths.antigravityMcpFiles()) {
          const cfg = (readJsonFile(f) as Record<string, unknown>) ?? {};
          const srv =
            (cfg.mcpServers as Record<string, { command?: string; args?: string[] }>) ?? {};
          const entry = srv[tool];
          if (!entry)
            return { healthy: false, reason: "not configured in all antigravity targets" };
          const h = mcpSpawnHealthy(entry.command ?? "", entry.args ?? []);
          if (!h.healthy) return h;
        }
        return { healthy: true };
      }
    }
  } catch (err) {
    return { healthy: false, reason: err instanceof Error ? err.message : String(err) };
  }
}

/** Run the doctor command: health check. */
export async function run(offline: boolean, fix: boolean, opts: RunOpts): Promise<number> {
  colors.banner("toksave doctor", "quick health check");

  const PAD = Math.max(
    18,
    ...ALL_AGENTS.map((a) => a.label.length + 2),
    ...ALL_TOOLS.map((t) => t.label.length + 2),
  );

  // ── Per-agent wiring status ─────────────────────────────
  for (const agent of ALL_AGENTS) {
    const det = detectAgent(agent.id);
    if (!det.installed) {
      colors.raw(
        `  ${pc.dim(colors.BULLET)} ${colors.pad(agent.label, PAD)}${pc.dim("not installed")}`,
      );
      continue;
    }

    const missing: string[] = [];
    for (const tool of ALL_TOOLS) {
      const verified = verifyTool(agent.id, tool.id);
      if (verified === false || verified === null) {
        missing.push(tool.label);
      } else if (tool.id === "codegraph" || tool.id === "context-mode") {
        // Deep verification: presence in config is not enough if binary won't launch
        const mcpStatus = checkAgentMcpHealth(agent.id, tool.id);
        if (!mcpStatus.healthy) {
          missing.push(`${tool.label} (${mcpStatus.reason ?? "unexecutable"})`);
        }
      }
    }

    if (missing.length === 0) {
      colors.raw(
        `  ${pc.green(colors.CHECK)} ${colors.pad(agent.label, PAD)}${pc.dim("all tools wired")}`,
      );
    } else {
      colors.raw(
        `  ${pc.yellow(colors.WARN)} ${colors.pad(agent.label, PAD)}${pc.yellow(`missing: ${missing.join(", ")}`)}`,
      );
    }
  }

  // ── Tool versions ───────────────────────────────────────
  if (!offline) {
    console.log();
    let outdated = 0;

    for (const tool of ALL_TOOLS) {
      const installed = toolInstalledVersion(tool.id);
      const latest = await toolLatestVersion(tool.id);
      const label = colors.pad(tool.label, PAD);

      if (tool.instructionOnly) {
        colors.raw(`  ${pc.green(colors.CHECK)} ${pc.dim(label)}${pc.dim("instruction-only")}`);
      } else if (installed && latest) {
        const instStr = installed.startsWith("v") ? installed : `v${installed}`;
        const latestStr = latest.startsWith("v") ? latest : `v${latest}`;
        if (isUpToDate(installed, latest)) {
          colors.raw(`  ${pc.green(colors.CHECK)} ${pc.dim(label)}${pc.dim(instStr)}`);
        } else {
          outdated++;
          colors.raw(
            `  ${pc.yellow(colors.ARROW_UP)} ${pc.dim(`${label}${instStr}`)}${pc.green(` → ${latestStr}`)}`,
          );
        }
      } else if (installed) {
        const instStr =
          installed === "skill" || installed === "instruction-only"
            ? installed
            : installed.startsWith("v")
              ? installed
              : `v${installed}`;
        colors.raw(`  ${pc.green(colors.CHECK)} ${pc.dim(label)}${pc.dim(instStr)}`);
      } else {
        colors.raw(`  ${pc.dim(colors.BULLET)} ${pc.dim(label)}${pc.dim("not installed")}`);
      }
    }

    console.log();
    if (outdated > 0) {
      colors.warn(`${outdated} update(s) available — run \`toksave update\``);
    } else {
      colors.ok("All up to date.");
    }
  }

  // ── PATH auto-fix ────────────────────────────────────────
  if (fix) {
    const msg = formatPathFixResult(selfHealPath());
    if (msg) colors.ok(msg);
  }

  // ── Tool health ─────────────────────────────────────────
  const unhealthy = ALL_TOOLS.map((tool) => ({ tool, health: toolHealthCheck(tool.id) })).filter(
    ({ health }) => !health.healthy,
  );

  if (unhealthy.length > 0) {
    console.log();
    for (const { tool, health } of unhealthy) {
      const label = colors.pad(tool.label, PAD);
      colors.raw(`  ${pc.yellow(colors.WARN)} ${label}${pc.yellow("unhealthy")}`);
      printHealthIssues(health);

      if (fix) {
        const result = await toolRepair(tool.id, opts);
        const icon = result.success ? pc.green(colors.CHECK) : pc.red(colors.CROSS);
        colors.raw(`  ${icon} ${colors.pad(tool.label, PAD)}${result.message}`);
        if (result.healthAfterRepair) {
          const status = result.healthAfterRepair.healthy
            ? pc.green("healthy")
            : pc.yellow("unhealthy");
          colors.raw(
            `  ${pc.dim(colors.BULLET)} ${colors.pad(tool.label, PAD)}after repair: ${status}`,
          );
          printHealthIssues(result.healthAfterRepair);
        }
      }
    }

    if (!fix) {
      console.log();
      colors.info("Run `toksave doctor --fix` to repair unhealthy tools.");
    }
  }

  // ── Fix suggestion ──────────────────────────────────────
  const broken = ALL_AGENTS.some((a) => {
    const det = detectAgent(a.id);
    if (!det.installed) return false;
    return ALL_TOOLS.some((t) => verifyTool(a.id, t.id) === false);
  });

  if (broken) {
    console.log();
    colors.info("Run `toksave` to fix.");
  }

  printRepoFooter();
  console.log();
  return 0;
}

function printRepoFooter(): void {
  if (process.env.TOKSAVE_TEST === "1") return;
  const repoURL = "https://github.com/agungprasastia/toksave";
  const issuesURL = `${repoURL}/issues`;
  console.log();
  console.log(`  ${pc.dim("─".repeat(52))}`);
  console.log(`  ${pc.dim("If toksave helps, please star it here: ")}${pc.cyan(repoURL)}`);
  console.log(
    `  ${pc.dim("If you hit any issue or have ideas, please raise it here: ")}${pc.cyan(issuesURL)}`,
  );
}

function printHealthIssues(health: HealthStatus): void {
  for (const issue of health.issues) {
    const icon = issue.severity === "error" ? pc.red(colors.CROSS) : pc.yellow(colors.WARN);
    colors.raw(`    ${icon} ${issue.message}`);
    if (issue.remediation) colors.raw(`      ${pc.dim(issue.remediation)}`);
  }
}

File: src/commands/doctor.rs
use crate::cli::ParsedCli;
use crate::registry::{detect_agent, verify_tool, ALL_AGENTS, ALL_TOOLS};
use crate::tools::{tool_health_check, tool_installed_version, tool_latest_version, tool_repair};
use crate::util::colors;
use crate::util::health::{HealthStatus, Severity};
use crate::util::probe::probe_agent;
use colored::Colorize;

pub async fn run_doctor(parsed: &ParsedCli, offline: bool, fix: bool) -> i32 {
    colors::banner("toksave doctor", "quick health check");

    let pad = ALL_AGENTS
        .iter()
        .map(|a| a.label.len())
        .chain(ALL_TOOLS.iter().map(|t| t.label.len()))
        .max()
        .unwrap_or(16)
        .max(16)
        + 2;

    // Per-agent wiring status
    for agent in ALL_AGENTS {
        let det = detect_agent(agent.id);
        let label = format!("{:<width$}", agent.label, width = pad);
        if !det.installed {
            println!(
                "  {} {}{}",
                colors::BULLET.dimmed(),
                label,
                "not installed".dimmed()
            );
            continue;
        }
        let mut missing: Vec<String> = vec![];
        for tool in ALL_TOOLS {
            match verify_tool(agent.id, tool.id) {
                Some(true) => {}
                _ => missing.push(tool.label.to_string()),
            }
        }
        if missing.is_empty() {
            println!(
                "  {} {}{}",
                colors::CHECK.green(),
                label,
                "all tools wired".dimmed()
            );
        } else {
            let missing_str = format!("missing: {}", missing.join(", "));
            println!(
                "  {} {}{}",
                colors::WARN.yellow(),
                label,
                missing_str.yellow()
            );
        }

        // Runtime probe: wired hook/MCP commands must resolve and run.
        for issue in probe_agent(agent.id) {
            println!(
                "      {} {}{}",
                colors::WARN.yellow(),
                issue.kind,
                format!(" — {}", issue.detail).yellow()
            );
        }
    }

    // Tool versions (skip when offline)
    if !offline {
        println!();
        let mut outdated = 0usize;
        for tool in ALL_TOOLS {
            let installed = tool_installed_version(tool.id);
            let latest = tool_latest_version(tool.id).await;
            let label = format!("{:<width$}", tool.label, width = pad);
            if tool.instruction_only {
                println!(
                    "  {} {}{}",
                    colors::CHECK.green(),
                    label.dimmed(),
                    "instruction-only".dimmed()
                );
            } else if let Some(inst) = installed {
                let inst_str = if inst.starts_with('v') {
                    inst.clone()
                } else {
                    format!("v{inst}")
                };
                match latest {
                    Some(lat) if !version_up_to_date(&inst, &lat) => {
                        outdated += 1;
                        let lat_str = if lat.starts_with('v') {
                            lat.clone()
                        } else {
                            format!("v{lat}")
                        };
                        println!(
                            "  {} {}{}",
                            "↑ ".yellow(),
                            format!("{label}{inst_str}").dimmed(),
                            format!(" → {lat_str}").green()
                        );
                    }
                    _ => println!(
                        "  {} {}{}",
                        colors::CHECK.green(),
                        label.dimmed(),
                        inst_str.dimmed()
                    ),
                }
            } else {
                println!(
                    "  {} {}{}",
                    colors::BULLET.dimmed(),
                    label.dimmed(),
                    "not installed".dimmed()
                );
            }
        }
        println!();
        if outdated > 0 {
            colors::warn(&format!(
                "{outdated} update(s) available — run `toksave update`"
            ));
        } else {
            colors::ok("All up to date.");
        }
    }

    // Tool health
    let unhealthy: Vec<_> = ALL_TOOLS
        .iter()
        .filter_map(|t| {
            let h = tool_health_check(t.id);
            if h.healthy {
                None
            } else {
                Some((t, h))
            }
        })
        .collect();

    if !unhealthy.is_empty() {
        println!();
        for (tool, health) in &unhealthy {
            let label = format!("{:<width$}", tool.label, width = pad);
            println!(
                "  {} {}{}",
                colors::WARN.yellow(),
                label,
                "unhealthy".yellow()
            );
            print_health_issues(health);
            if fix {
                let result = tool_repair(tool.id, &parsed.opts).await;
                let icon = if result.success {
                    colors::CHECK.green()
                } else {
                    colors::CROSS.red()
                };
                let repair_label = format!("{:<width$}", tool.label, width = pad);
                println!("  {} {}{}", icon, repair_label, result.message);
                if let Some(after) = &result.health_after_repair {
                    let status = if after.healthy {
                        "healthy".green()
                    } else {
                        "unhealthy".yellow()
                    };
                    let after_label = format!("{:<width$}", tool.label, width = pad);
                    println!(
                        "  {} {}after repair: {}",
                        colors::BULLET.dimmed(),
                        after_label,
                        status
                    );
                    print_health_issues(after);
                }
            }
        }
        if !fix {
            println!();
            colors::info("Run `toksave doctor --fix` to repair unhealthy tools.");
        }
    }

    // Suggest init if any installed agent has unwired tools
    let broken = ALL_AGENTS.iter().any(|a| {
        detect_agent(a.id).installed
            && ALL_TOOLS
                .iter()
                .any(|t| verify_tool(a.id, t.id) == Some(false))
    });
    if broken {
        println!();
        colors::info("Run `toksave` to fix.");
    }

    println!();
    0
}

fn print_health_issues(health: &HealthStatus) {
    for issue in &health.issues {
        let icon = match issue.severity {
            Severity::Error => colors::CROSS.red(),
            Severity::Warning => colors::WARN.yellow(),
        };
        println!("    {} {}", icon, issue.message);
        if let Some(rem) = &issue.remediation {
            println!("      {}", rem.dimmed());
        }
    }
}

fn version_up_to_date(installed: &str, latest: &str) -> bool {
    let parse = |s: &str| -> Vec<u64> {
        s.trim_start_matches('v')
            .split('.')
            .map(|p| {
                p.chars()
                    .take_while(|c| c.is_ascii_digit())
                    .collect::<String>()
            })
            .map(|p| p.parse().unwrap_or(0))
            .collect()
    };
    let i = parse(installed);
    let l = parse(latest);
    for k in 0..3 {
        let iv = i.get(k).copied().unwrap_or(0);
        let lv = l.get(k).copied().unwrap_or(0);
        if iv != lv {
            return iv > lv;
        }
    }
    true
}
